inmoose/sim/splat.py

```diff
@@ -0,0 +1,195 @@
+        the biological coefficient of variation parameters
+    x0 : float, optional
+        dropout midpoint, used to draw the dropouts. Unused if
+        `single_cell=False`.
+    k : float, optional
+        dropout shape, used to draw the dropouts. Unused if `single_cell=False`.
+    random_state : {int, `numpy.random.Generator` or `numpy.random.RandomState`},
+                    optional
+        If `random_state` is None (or `np.random`), the
+        `numpy.random.RandomState` singleton is used.  If `random_state` is an
+        int, a new ``RandomState`` instance is used, seeded with
+        `random_state`.  If `random_state` is already a ``Generator`` or
+        ``RandomState`` instance, that instance is used.
+
+    Returns
+    -------
+    ndarray
+        simulated count matrix, of size `nb_genes` x `nb_samples`
+    """
+
+    # make sure we have a Generator
+    random_state = np.random.default_rng(random_state)
+
+    if batch is None:
+        batch = np.zeros(nb_samples, dtype=np.int64)
+    batch = np.asarray(batch)
+    assert len(batch) == nb_samples
+
+    if group is None:
+        group = np.zeros(nb_samples, dtype=np.int64)
+    group = np.asarray(group)
+    assert len(group) == nb_samples
+
+    nb_batches = len(np.unique(batch))
+    nb_groups = len(np.unique(group))
+
+    # draw gene means
+    original_means = gamma.rvs(alpha, beta, size=nb_genes, random_state=random_state)
+
+    # draw outliers
+    is_outlier = bernoulli.rvs(outlier_prob, size=original_means.shape, random_state=random_state)
+    outlier_factor = lognorm.rvs(outlier_location, outlier_scale, size=original_means.shape, random_state=random_state)
+    gene_means = np.where(is_outlier, outlier_factor * np.median(original_means), original_means)
+
+    # draw library sizes
+    L = lognorm.rvs(libsize_loc, libsize_scale, size=nb_samples, random_state=random_state)
+
+    # cell means
+    cell_means = L * gene_means[:,None] / gene_means.sum()
+
+    # biological coefficients of variation (squared)
+    B2 = (phi + 1/np.sqrt(cell_means))**2 * (bcv_df / chi2.rvs(bcv_df, size=nb_genes, random_state=random_state))[:,None]
+
+    # trended cell means
+    trended_cell_means = gamma.rvs(1/B2, cell_means * B2, random_state=random_state)
+
+    # DE fold-changes between cell types
+    de_lf = [get_lognorm_factors(nb_genes, 1, 0.3+0.2*i, 0.1 + i*0.5, 1, random_state=random_state) for i in range(nb_groups)]
+    de_lf = np.choose(group, [x[:,None] for x in de_lf])
+
+    # cell type specific cell means
+    de_trended_cell_means = de_lf * trended_cell_means
+
+    # batch effect factors
+    be_fact = [get_lognorm_factors(nb_genes, 1, 0.5, 0.1 + 0.5*i, 1, random_state=random_state) for i in range(nb_batches)]
+    be_fact = np.choose(batch, [x[:,None] for x in be_fact])
+    batch_cell_means = be_fact * de_trended_cell_means
+
+    # counts
+    y = poisson.rvs(batch_cell_means, random_state=random_state)
```

Comment:
We saw that with Kader, is this model creating Zero Inflated Poisson distribution instead on Negative Binomial?

Reply:
Yes and no ;-)
The Splat model creates a zero-inflated Poisson distribution, whose parameter follows a Gamma distribution (modified to account for outliers, batches...). Such Poisson-Gamma mixture is in fact a negative binomial distribution (see [the maths here](https://en.wikipedia.org/wiki/Negative_binomial_distribution#Gamma%E2%80%93Poisson_mixture)). I would say it is another view on the same mathematical object.

It just so happens that outliers, batches, mean-variance dependency... are easier to model in the Poisson-Gamma view.